packages/belos/src/BelosGmresPolyOp.hpp

```diff
@@ -606,23 +652,247 @@ namespace Belos {
+    std::vector<MagnitudeType> pof (dim_,1.0);
+    for(int j=0; j<dim_; ++j) {
+      for(int i=0; i<dim_; ++i) {
+        if(i!=j) {
+          pof[j] = std::abs(pof[j]*(1.0-(cmplxHRitz[j]/cmplxHRitz[i])));
+        }
+      }
+    }
+
+    // Compute number of extra roots needed:
+    std::vector<int> extra (dim_);
+    int totalExtra = 0;
+    for(int i=0; i<dim_; ++i){
+      extra[i] = ceil((log10(pof[i])-4.0)/14.0);
+      if(extra[i] > 0){
+        totalExtra += extra[i];
+      }
+    }
+    if (totalExtra){
+      printer_->stream(Warnings) << "Warning: Need to add " << totalExtra << " extra roots." << std::endl;}
+
+    // If requested to add roots, append them to the theta matrix:
+    if(addRoots_ && totalExtra>0)
+    {
+      theta_.reshape(dim_+totalExtra,2);
+      // Make a matrix copy for perturbed roots:
+      Teuchos::SerialDenseMatrix<OT,MagnitudeType> thetaPert (Teuchos::Copy, theta_, dim_+totalExtra, 2);
+
+      //Add extra eigenvalues to matrix and perturb for sort:
+      int count = dim_;
+      for(int i=0; i<dim_; ++i){
+        for(int j=0; j< extra[i]; ++j){
+          theta_(count,0) = theta_(i,0);
+          theta_(count,1) = theta_(i,1);
+          thetaPert(count,0) = theta_(i,0)+(j+MCT::one())*5e-8;
+          thetaPert(count,1) = theta_(i,1);
+          ++count;
+        }
+      }
+
+      // Update polynomial degree:
+      dim_ += totalExtra;
+      if (totalExtra){
+        printer_->stream(Warnings) << "New poly degree is: " << dim_ << std::endl;}
+
+      // Create a new index and sort perturbed roots:
+      std::vector<int> index2(dim_);
+      for(int i=0; i<dim_; ++i){ 
+        index2[i] = i; 
+      }
+      SortModLeja(thetaPert,index2);
+      //Apply sorting to non-perterbed roots:
+      for(int i=0; i<dim_; ++i)
+      { 
+        thetaPert(i,0) = theta_(index2[i],0);
+        thetaPert(i,1) = theta_(index2[i],1);
+      }
+      theta_ = thetaPert;
+
+    } //----End root adding section------
+
+   }//End section for creating Roots poly 
   }
 
+  template <class ScalarType, class MV, class OP>
+  void GmresPolyOp<ScalarType, MV, OP>::SortModLeja(Teuchos::SerialDenseMatrix< OT, MagnitudeType > &thetaN, std::vector<int> &index) const 
+  {
+    //Sort theta values via Modified Leja Ordering:
+
+    // Set up blank matrices to track sorting:
+    int dimN = index.size();
+    std::vector<int> newIndex(dimN);
+    Teuchos::SerialDenseMatrix< OT, MagnitudeType > sorted (thetaN.numRows(), thetaN.numCols());
+    Teuchos::SerialDenseMatrix< OT, MagnitudeType > absVal (thetaN.numRows(), 1);
+    Teuchos::SerialDenseMatrix< OT, MagnitudeType > prod (thetaN.numRows(), 1);
+
+    //Compute all absolute values and find maximum:
+    for(int i = 0; i < dimN; i++){
+      absVal(i,0) = sqrt(thetaN(i,0)*thetaN(i,0) + thetaN(i,1)*thetaN(i,1));
+    }
+    MagnitudeType * maxPointer = std::max_element(absVal.values(), (absVal.values()+dimN));
+    int maxIndex = (maxPointer- absVal.values());
```

Comment:
This is a narrowing conversion and may thus introduce a build warning.  The proper type is `ptrdiff_t`.  You could always cast the right-hand side explicitly to `int` to avoid possible warnings.

Reply:
I type casted to int on the right-hand side.